src/entities.cs

```diff
@@ -58,6 +83,28 @@ public struct EcsPackedEntityWithWorld {
             return $"Entity-{Id}:{Gen} [{sb}]";
         }
 #endif
+        public bool Equals(EcsPackedEntityWithWorld other)
+        {
+            return Id == other.Id && Gen == other.Gen && Equals(World, other.World);
```

Comment:
Ideally we would use this new IEquatable.Equals in
```cs
public static bool EqualsTo (this in EcsPackedEntityWithWorld a, in EcsPackedEntityWithWorld b) {
    return a.Id == b.Id && a.Gen == b.Gen && a.World == b.World;
}
```
To make sure the API is consistent & avoid redundancy.

But I'm wondering if there are breaking scenarios between == and Equals. Because == will compare through reference, while equals will check for equality/box?

Reply:
Ooh, good catch, that should be updated too!

I'm not an expert on structs, but my understanding is that, by default ,  struct equality is value equality through a .Equals implementation that uses Reflection to check that all the fields are equal, see technical document above and https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1815.  However, there is no '==' operator defined for structs by default.  Exactly *why* the documentation recommends specifying the == and != operators for structs when specializing the .Equals is something I'm not clear on.  The operator is not valid on structs unless it has been defined, so it would seem to be a safe omission.  Microsoft even states this here: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/how-to-define-value-equality-for-a-type

> For structs, the default implementation of [Object.Equals(Object)](https://docs.microsoft.com/en-us/dotnet/api/system.object.equals#system-object-equals(system-object)) (which is the overridden version in [System.ValueType](https://docs.microsoft.com/en-us/dotnet/api/system.valuetype)) performs a value equality check by using reflection to compare the values of every field in the type. When an implementer overrides the virtual Equals method in a struct, the purpose is to provide a more efficient means of performing the value equality check and optionally to base the comparison on some subset of the struct's field or properties.
>
> The [==](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/equality-operators#equality-operator-) and [!=](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/equality-operators#inequality-operator-) operators can't operate on a struct unless the struct explicitly overloads them.

I'm just following Microsoft's recommendations here on specifying it when adding IEquatable, and I can't think of a reason to disallow the == operator.

I think we're safe from breaking changes, because == was never allowed on EcsPackedEntity or EcsPackedEntityWithWorld before, due to the above.
![image](https://user-images.githubusercontent.com/33366270/168604768-295b0271-2e98-453d-b015-f9486657661a.png)


And so in all cases it SHOULD have been doing value equality before, just in a boxing, reflection-based way.

Reply:
Oh, this is my bad. I meant the world == check. Forgot to add that information

Reply:
OH!

EcsWorld is a class, so there's no boxing there, and similarly both ways of writing it would do a reference check, since there is no specialized .Equals or == operator on EcsWorld.  Backwards compatibility should be fine there.


However, with an eye looking forward and which form to favor, it all comes down to expectation with regards to deriving from EcsWorld

1) `==` for object types explicitly performs a ReferenceEquals() check, if no specialized operator exists
2) operators are statically determined at runtime, whereas Equals is runtime-determined based on the types being compared (that's actually how boxed value-types still check value equality).
3) operators are NOT inherited, whereas .Equals is polymorphic.
4) EcsWorld is NOT a sealed class, and could be derived from  (I'd argue *why?*, but I'm not going to pretend that I understand every use case).

== will work fine (and be micro-faster due to static compilation) so long as users stick to the default EcsWorld implementation or want reference equality checks, regardless.  Putting == in the EcsPackedEntityWithWorld equality would prevent any specialized equality detection and just do simple reference equality checks, even if they specified the == operator, due to the static compilation.  Equals respects polymorphism, so avoids this issue.   (and the Equals(obj, obj) form just handles null, unlike (obj.Equals(obj))

I'm fine changing the code either way.  == is a bit faster, but it will force reference equality on derived classes.  I'm still okay with that as I have yet to encounter any case were a) I would want to derive from EcsWorld, b) I wouldn't want reference equality, but it does leave the door open to a potential (extremely unlikely) issue later.  Another alternative is to make EcsWorld a sealed class and just slam the door on the whole problem.

Reply:
Mhm, was just making sure that these niche scenarios where people do indeed ex. inherit EcsWorld still work the same with LeoEcsCommunity code. Honestly, I'd say we leave it with the old == operator. And change it maybe at some point to Equals, if needed. Better to leave old functionality as it is, for the niche scenarios we might not be aware of right now.

Reply:
Okay, updated!